espnet/nets/pytorch_backend/e2e_asr_transducer.py

```diff
@@ -0,0 +1,414 @@
+#!/usr/bin/env python
+
+import argparse
+import logging
+import math
+
+import editdistance
+
+import chainer
+import numpy as np
+import six
+import torch
+
+from chainer import reporter
+
+from espnet.nets.asr_interface import ASRInterface
+from espnet.nets.pytorch_backend.rnn.attentions import att_for
+from espnet.nets.pytorch_backend.rnn.decoders_transducer import decoder_for
+from espnet.nets.pytorch_backend.rnn.encoders import encoder_for
+
+from espnet.nets.pytorch_backend.nets_utils import pad_list
+from espnet.nets.pytorch_backend.nets_utils import to_device
+from espnet.nets.pytorch_backend.nets_utils import to_torch_tensor
+
+
+class Reporter(chainer.Chain):
+    """A chainer reporter wrapper"""
+
+    def report(self, loss, cer, wer):
+        reporter.report({'cer': cer}, self)
+        reporter.report({'wer': wer}, self)
+        logging.info('loss:' + str(loss))
+        reporter.report({'loss': loss}, self)
+
+
+class E2E(ASRInterface, torch.nn.Module):
+    """E2E module
+
+    Args:
+        idim (int): dimension of inputs
+        odim (int): dimension of outputs
+        args (namespace): argument Namespace containing options
+    """
+
+    @staticmethod
+    def add_arguments(parser):
+        group = parser.add_argument_group("transducer model setting")
+        group.add_argument('--rnnt_type', default='warp-transducer', type=str,
+                           choices=['warp-transducer'],
+                           help='Type of transducer implementation to calculate loss.')
+        parser.add_argument('--rnnt-mode', default=0, type=int, choices=[0, 1],
+                            help='RNN-Transducing mode (0:rnnt, 1:rnnt-att)')
```

Comment:
`'--rnnt-mode', default=0, type=str, choices=['rnnt', 'rnnt-att'],` is more expressive